### ISSUE TITLE
[Feature] support `--modules-folder`

### ISSUE THREAD

**Issue by nektro** [yarnpkg/berry#5441]
- [ ] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [ ] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**

As a developer that regularly stores my code repositories on storage media such as disk HDDs or small flash media like SD cards or USB drives the amount of files that `node_modules` creates can create a lot of otherwise unnecessary stress on these types of storage and not only be slow but hurt the lifetimes. Being able to specify a separate path on more apt media like an SSD would significantly speed up my development and help my drives last longer.

**Describe the solution you'd like**

Yarn v1 had this flag.

**Describe the drawbacks of your solution**

.

**Describe alternatives you've considered**

.

**Comment by dlqqq** [comment]
I have another use case that will hopefully bring more attention to this issue: our package needs two different sets of dependencies because our package needs to work with two different major versions of JupyterLab. It would be really nice to have separate `node_modules` directories so I don't have to redownload all the dependencies every time I switch between versions in local development.

**Comment by RobMayer** [comment]
I also have a use-case for this, though it is very niche. I have an application that requires all dependencies to be placed into a specific folder to be used by a v8 runtime within that application. I used this to automate doing this for v1, and I cannot start using berry for this project until this gets added.

**Comment by ff0041** [comment]
I used the `--modules-folder` feature in yarn classic to install dependencies into a shared Docker volume, which VSCode accesses from a devcontainer, allowing features like `Go to Definition` to work. Might need a different approach for yarn berry...

**Comment by JaneSmith** [comment]
My use case for this is with dev containers in VS Code. I want the modules to go into the container, sharing the lifecycle of the container. When I delete the container, the modules get deleted with it. No clutter in my project directory.

**Comment by BasixKOR** [comment]
I don't think using the same directories for multiple project is going to provide a stable and consistent module resolution, so I think the better approach to tackle this issue would be enable some feature Yarn had for quite a while now.

You can enable `enableGlobalCache` and `nmMode: hardlinks-global` to enable further space optimizations, reducing the disk usage a lot if you have multiple projects.

**Comment by RobMayer** [comment]
> I don't think using the same directories for multiple project is going to provide a stable and consistent module resolution, so I think the better approach to tackle this issue would be enable some feature Yarn had for quite a while now.
> 
> You can enable `enableGlobalCache` and `nmMode: hardlinks-global` to enable further space optimizations, reducing the disk usage a lot if you have multiple projects.

I'm not sure I follow. I don't see anyone citing mulitple projects using the same modules folder as a use-case. who"a problem is this advice intended to help solve?

**Comment by BasixKOR** [comment]
@RobMayer My bad, I got confused for a moment lol. But the latter advice still stands; there are options you can enable to save space without changing directories.

Now I understand the issue better, something I'm missing is that: doesn't moving node_modules break resolutions? Like, if you install dependencies of `~/project/A/` on `~/cache/node_modules` or somewhere else, Node.js instance running on `~/project/A` wouldn't be able to find the modules. `yarn run` and other essential commands won't work in that case.

**Comment by RobMayer** [comment]
That's a good question - I don't know how this affects module resolution. For my own use-case it's not relevant: My own use-case is to essentially create a complete production-ready copy of node_modules in another directory for use in another another non-node environment. module resolution in the original project is unhampered.
